dashboard/app/models/user.rb

```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
       # if we have one or more sections matching this script id, we consider a stage hidden if all of those sections
       # hides the stage
-      script_sections.all? {|s| script_level.stage_hidden_for_section?(s.id)}
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extra a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  def get_hidden_ids(assign_id, hidden_stages = true)
+    # If we're a teacher, we want to go through each of our sections and return
+    # a mapping from section id to hidden stages in that section
+    if teacher?
+      hidden_by_section = {}
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
+      end
+      return hidden_by_section
+    end
+
+    # if we're a student, we want to look through each of the sections in which
+    # we're a member, and use those to figure out which stages should be hidden
+    # for us
+    sections = sections_as_student
+    return [] if sections.empty?
+
+    assigned_sections = sections.select do |section|
+      hidden_stages && section.script_id == assign_id || !hidden_stages && section.course_id == assign_id
+    end
+
+    if assigned_sections.empty?
+      # if we have no sections matching this assignment, we consider a stage/script
+      # hidden if any of our sections hides it
+      if hidden_stages
+        sections.flat_map(&:section_hidden_stages).pluck(:stage_id).uniq
+      else
+        sections.flat_map(&:section_hidden_scripts).pluck(:script_id).uniq
+      end
+    else
+      # if we do have sections matching this assignment, we consider a stage/script
+      # hidden only if it is hidden in every one of the sections the student belongs
+      # to that match this assignment
+      if hidden_stages
+        all_ids = assigned_sections.flat_map(&:section_hidden_stages).pluck(:stage_id)
+      else
+        all_ids = assigned_sections.flat_map(&:section_hidden_scripts).pluck(:script_id)
+      end
+
+      counts = all_ids.each_with_object(Hash.new(0)) {|id, hash| hash[id] += 1}
+      counts.select {|_, val| val == assigned_sections.length}.keys
+    end
+  end
```

Comment:
I was somewhat torn on how to best write this method.

I originally considered taking the initial `hidden_stage?` method, duplicating it, and then changing it to find scripts instead of stages as appropriate. This however was a lot of duplicated code.

I then explored passing in the map function (i.e. section_hidden_stages vs section_hidden_scripts), and other data (i.e. :stage_id vs. script_id) as params. I never got this fully working, but could have if I'd stuck with it. However, the resulting method felt very difficult to read.

I ultimately settled on this approach of having a bunch of places where we gate on `hidden_stages` or not. I don't love it, but I prefer it to the alternatives.